cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java

```diff
@@ -82,16 +87,36 @@ public void healthCheck(HttpServiceRequest request, HttpServiceResponder respond
   @Path("v1/contexts/{context}/validations/stage")
   public void validateStage(HttpServiceRequest request, HttpServiceResponder responder,
                             @PathParam("context") String namespace) throws IOException, AccessException {
-    validateLocally(request, responder, namespace);
-  }
-
-  private void validateLocally(HttpServiceRequest request, HttpServiceResponder responder,
-                               String namespace) throws IOException {
     if (!getContext().getAdmin().namespaceExists(namespace)) {
       responder.sendError(HttpURLConnection.HTTP_NOT_FOUND, String.format("Namespace '%s' does not exist", namespace));
       return;
     }
 
+    //Validate remotely if remote execution is enabled
+    if (getContext().remoteExecutionEnabled()) {
+      validateRemotely(request, responder, namespace);
+      return;
+    }
+    validateLocally(request, responder, namespace);
+  }
+
+  private void validateRemotely(HttpServiceRequest request, HttpServiceResponder responder,
+                                String namespace) throws IOException {
+    String validationRequestString = StandardCharsets.UTF_8.decode(request.getContent()).toString();
+    RemoteValidationRequest remoteValidationRequest = new RemoteValidationRequest(namespace, validationRequestString);
+    RunnableTaskRequest runnableTaskRequest = RunnableTaskRequest.getBuilder(RemoteValidationTask.class.getName()).
+      withParam(GSON.toJson(remoteValidationRequest)).
+      build();
+    try {
+      byte[] bytes = getContext().runTask(runnableTaskRequest);
+      responder.sendString(Bytes.toString(bytes));
+    } catch (Exception e) {
+      responder.sendError(HttpURLConnection.HTTP_INTERNAL_ERROR, "Error from remote task " + e.getMessage());
```

Comment:
Is always returning 500 on error preserving the current non-remote validation behavior?

Reply:
Moved RemoteExecutionException, RemoteTaskException to system-app-api to better handle the exceptions and return correct error code.